compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUTile.cpp

```diff
@@ -0,0 +1,120 @@
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Linalg/Utils/Utils.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/MemRef/Transforms/Passes.h"
+#include "mlir/Dialect/SCF/Transforms/TileUsingInterface.h"
+#include "mlir/Dialect/SCF/Transforms/Transforms.h"
+#include "mlir/IR/Iterators.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-llvmcpu-tile"
+
+namespace mlir {
+namespace iree_compiler {
+namespace {
+
+/// Builds a proper tile sizes vector for the op.
+/// scf::tileUsingSCFForOp expects the num of tile sizes = num of loops. This
+/// method returns a proper tile sizes vector for each op during tiling.
+static SmallVector<Value> buildTileSizesForOp(OpBuilder &b, Operation *op,
+                                              ArrayRef<int64_t> tileSizes) {
+  auto tilingOp = cast<TilingInterface>(op);
+
+  SmallVector<int64_t> newTileSizes(tileSizes);
+  newTileSizes.resize(tilingOp.getLoopIteratorTypes().size(), /*default=*/0);
+
+  OpBuilder::InsertionGuard guard(b);
+  return llvm::to_vector(map_range(newTileSizes, [&](int64_t size) {
+    Value v = b.create<arith::ConstantIndexOp>(tilingOp->getLoc(), size);
+    return v;
+  }));
+}
+
+/// This pass tiles all the TilingInterface operations. The `tilingLevel` must
+/// be specified. It picks the `tilingLevel`-th list as tiling sizes from
+/// lowering_config.
+struct LLVMCPUTilePass : LLVMCPUTileBase<LLVMCPUTilePass> {
+  LLVMCPUTilePass(int64_t tilingLevel = -1) {
+    this->tilingLevel.setValue(tilingLevel);
+  }
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<arith::ArithDialect, AffineDialect, linalg::LinalgDialect,
+                    scf::SCFDialect>();
+  }
+
+  void runOnOperation() override;
+};
+
+void LLVMCPUTilePass::runOnOperation() {
+  if (tilingLevel == -1) {
+    LLVM_DEBUG(llvm::dbgs() << "tilingLevel not set, skip tiling");
+    return;
+  }
+  MLIRContext *context = &getContext();
+  auto funcOp = getOperation();
+
+  SmallVector<Operation *> computeOps = getComputeOps(funcOp);
+  FailureOr<IREE::Codegen::LoweringConfigAttr> maybeLoweringConfig =
+      getLoweringConfig(computeOps);
+  if (failed(maybeLoweringConfig)) {
+    LLVM_DEBUG(llvm::dbgs() << "can't find lowering_config, skip tiling");
+    return;
+  }
+  SmallVector<int64_t> tileSizes =
+      maybeLoweringConfig.value().getTileSizeVals(tilingLevel);
+
+  for (auto computeOp : computeOps) {
+    auto op = cast<TilingInterface>(computeOp);
+    if (op.getLoopIteratorTypes().empty()) continue;
+
+    // For now do not tile `tensor.pad` operations. The `tensor.pad`
+    // operations might be those introduced by the padding-based
+    // codegeneration strategy. Those are not meant to be tiled again.
+    // Need a better way for handling this, but this works for now.
+    if (isa<tensor::PadOp>(computeOp)) continue;
```

Comment:
@MaheshRavishankar this should cover your changes in pad_fusion branch. And yes, we don't tile pad ops that are created by DoubleTilingPadExpert.